### ISSUE TITLE
HaTeX fails to parse when `$` not used for math mode

### ISSUE THREAD

**Issue by 2016rshah** [Daniel-Diaz/HaTeX#77]
HaTeX parses dollars signs as math mode, which is fine most of the time. But they can also be used in environments where they don't refer to math mode, and when they are, HaTeX will return Nothing rather than a syntax tree for the file. 

For example:
```
\begin{lstlisting}
$ ls
$ cd directory
$ emacs file
\end{lstlisting}
```
or 
```

\begin{minted}{haskell}
(+) <$> (Just 2) <*> (Just 3)
\end{minted}
```
The problem is that it expects `$` to be a part of the syntax tree even though it's just a symbol. Any ideas how to address this edge case, either by changing how HaTeX parses or by changing something in the LaTeX file (without getting rid of the dollar sign entirely)?

**Comment by Daniel-Diaz** [comment]
Oh, well. Another reason why LaTeX is so hard to parse. Every non-escaped dollar makes HaTeX think you are indeed in the beginning of a math expression. Does escaping the dollar signs (with ``\$``) work for you?

**Comment by 2016rshah** [comment]
From what I understand, it works fine if you escape the dollar sign. But because I'm using the dollar signs inside environments meant to display code, they will display almost everything I type inside that environment (including the slash) rather than interpreting that code as LaTeX. 

I took a quick glance at [the parsing code](https://github.com/Daniel-Diaz/HaTeX/blob/master/Text/LaTeX/Base/Parser.hs#L74), would it help to change the order we parse in to look for environments before looking for math mode blocks?

**Comment by Daniel-Diaz** [comment]
I don't think that's going to solve the problem. The dollar will still be parsed as math inside the environment. What we can do is to change the environment parser, so that it parses verbatim the content of certain environments. This, however, is not a general solution (and I doubt there is a general solution).

**Comment by leftaroundabout** [comment]
I agree. I _can't_ be possible to handle this kind of environment-specific parsing in full generality, short of doing TeX macro resolution, because of Turing-completeness. Hard-coding special cases for `verbatim`, `lstlisting` and `minted` is definitely ugly, but probably the most reasonable thing to do.

**Comment by leftaroundabout** [comment]
Or... perhaps we could refactor the parser, so it can take an extra argument: a function that prevents certain environments from being parsed?

**Comment by 2016rshah** [comment]
Ahh, after taking a closer look at the code, I understand what you mean now. What would the code look like to actually implement a change like this?

**Comment by Daniel-Diaz** [comment]
I'll work on this.

**Comment by Daniel-Diaz** [comment]
I implemented this following @leftaroundabout idea in commit https://github.com/Daniel-Diaz/HaTeX/commit/f8f6351f5f234da4268db63c41f4066a5abb6f84. I added a sample for the parser tester based on the example given by @2016rshah.

In summary, in addition to ``parseLaTeX``, you have ``parseLaTeXWith``, a function that allows to feed a configuration parameter to the parser. Currently, this configuration parameter only has a field with a list of verbatim environments. The default configuration includes the ``verbatim`` environment only.

**Comment by Daniel-Diaz** [comment]
I close this ticket. Re-open if the issue is not fixed.

**Comment by 2016rshah** [comment]
Went to try it out and I ran into Issue #80

**Comment by 2016rshah** [comment]
This works perfectly, thank you!
